enhancements/cert-manager-proxy/http01_challenge_proxy.md

```diff
@@ -0,0 +1,216 @@
+
+1. Cert Manager initiates an HTTP01 challenge for the API endpoint (`api.cluster.example.com`).
+2. The HTTP01 challenge request is directed to the API VIP on port 80.
+3. The HTTP01 Challenge Proxy intercepts the traffic using `nftables` and redirects it to the proxy pod on port 8888.
+4. The proxy pod forwards the request to the OpenShift Ingress Router, which serves the challenge response from the Cert Manager challenge pod.
+5. The ACME CA validates the challenge and issues the certificate for the API endpoint.
+
+### Topology Considerations
+
+- **Standard Clusters**: The API VIP is hosted on control plane nodes. The proxy ensures that HTTP01 challenges are redirected to the OpenShift Ingress Routers.
+- **Compact Clusters**: The proxy handles scenarios where the API VIP node may or may not host an OpenShift Router, ensuring consistent challenge redirection.
+- **SNO (Single Node OpenShift)**: The proxy is not strictly required in this topology, as the API and wildcard FQDNs resolve to the same IP. However, it can still be deployed for consistency.
+
+#### Hypershift / Hosted Control Planes
+
+This enhancement does not directly apply to Hypershift deployments, as the API endpoint management in Hypershift differs from baremetal environments. However, the proxy's design could be adapted for similar use cases in Hypershift if needed.
+
+#### Standalone Clusters
+
+For standalone clusters, the proxy ensures that HTTP01 challenges for the API endpoint are redirected to the OpenShift Ingress Routers, regardless of whether the API VIP node hosts a router.
+
+#### Single-node Deployments or MicroShift
+
+In SNO or MicroShift deployments, the proxy is not strictly required, as the API and wildcard FQDNs resolve to the same IP. However, deploying the proxy ensures consistency and simplifies certificate management.
+
+## Test Plan
+
+1. **Unit Tests**: Validate the proxy's functionality in isolation, including traffic redirection and error handling.
+2. **Integration Tests**: Deploy the proxy in a test cluster and verify that HTTP01 challenges for the API endpoint succeed.
+3. **Performance Tests**: Measure the proxy's impact on cluster performance and resource usage.
+4. **Topology Tests**: Test the proxy in Standard Clusters, Compact Clusters, and SNO environments to ensure compatibility.
+
+## Graduation Criteria
+
+### Dev Preview -> Tech Preview
+
+- The proxy is implemented and tested in development environments.
+- Documentation is available for deploying and configuring the proxy.
+
+### Tech Preview -> GA
+
+- The proxy is deployed in production environments and successfully handles HTTP01 challenges for various OCP topologies.
+- Performance and reliability meet production-grade requirements.
+
+### Removing a deprecated feature
+
+This enhancement does not deprecate any existing features.
+
+## Upgrade / Downgrade Strategy
+
+Updated versions of the proxy can be applied to the cluster similar to initial deployment
```

Comment:
Beyond the proxy itself, how do you plan to handle typical upgrade situations where there may older and newer instances of the proxy running at the same time due to a rolling upgrade? Are there any special considerations needed for this scenario?

What happens if an upgrade fails midway through? What steps would a user need to take to get their cluster back to the previous state?

Reply:
Adding additional context here.